### DISCUSSION TITLE
Route Guards for Tanstack Router

### DISCUSSION THREAD

**Discussion by PointSingularity** [TanStack/router#358] [Ideas]
Thanks for the great library, I really love what this router brings to the table (typesafety 👀)!

One thing that I would like to see is something akin to Guards in Angular/Vue.

Angular provides some nice [out-of-the-box](https://angular.io/api/router#structures) structures for route guards.
Custom guards are easy to make.

```typescript
@NgModule({
  imports: [
    RouterModule.forRoot([
      {
        path: 'team/:id',
        component: TeamComponent,
        canActivate: [CanActivateRoute] 
      }
    ])
  ],
  providers: [CanActivateRoute]
})
class AppModule {}
```

Some of the standouts are:
- CanActivate - which doesn't allow for loading an element based on some conditions (great for permissions), and won't load/trigger any elements or API calls for that route 
- CanDeactivate - can the route be navigated from, great for an agnostic way to show a confirmation dialog for all elements with a modifed unsaved form
...

Vue also provides some nice [features](https://router.vuejs.org/guide/advanced/navigation-guards.html): 

- router.beforeEach - applies to every route before load
- router.afterEach - applies to every route after load
...

The syntax is also pretty intuitive:
```typescript
const router = createRouter({ ... })

router.beforeEach((to, from) => {
  // ...
  // explicitly return false to cancel the navigation
  return false
})

```

Loaders can achieve a lot of these (maybe with some potential [issues](https://github.com/TanStack/router/discussions/92#discussioncomment-2447697)), but IMO it would be great to have.

**Comment by tannerlinsley** [comment]
The routing middle ware I’m doing soon will probably solve this

Tanner Linsley

**Reply by PointSingularity** [reply]
That is great to hear, thanks!

**Reply by starxyn** [reply]
Hi @tannerlinsley , can I check is the route guard implemented?
